### ISSUE TITLE
alternative custom opfor tables (AtB)

### ISSUE THREAD

**Issue by bczretribution** [MegaMek/mekhq#743]
[atbconfig(BCZ_LD).zip](https://github.com/MegaMek/mekhq/files/2078206/atbconfig.BCZ_LD.zip)
We have created a new set of opfor tables for AtB scenario generation.  You can just replace your current atbconfig.xml file with it.  It is currently only balanced for IS vs IS, we will be adding more for clan and CS later.  Overall we increased the variety of possible opfors for all weight classes,  added a small chance for an "easy" fight, and reduced the chance of an "overwhelming" fight but increased its difficulty.  This last change is to make it so that sometimes you need reinforcements to have a chance of victory (and sometimes you might just have to retreat).  There is also the occasionally damaged lance or reinforced lance.  We would like to hear any insights you have but in particular, it would be useful to hear which fights you think are too easy or too hard and the overall difficulty of it.

edit:  there seems to be an issue where mhq is not generating from the tables appropriately,  we are working on resolving the issue.  

Further investigation has shown that the these tables will not work without modification of the java.

**Comment by bczretribution** [comment]
So I talked at length with neo-ancient and the issue is that when the java reads from botlance it only expects to see "L" "UL" "M" "H" "A" "C" or "SH".  as you've found the java then turns each of those lances into a number 0-3 with some of them being aggregated together.  any other table entries in botlance the java doesn't know how to read.  in our file we completely redefined the table such that for example a traditional light lance (previously L=LLLM or LLLL) can instead be defined as A=LL and B=LL or LM so that AB=L, this methodology gave us the ability to have a much higher level of control over the strength and composition of the opfor.  so the issue is that the java is not flexible or robust enough to handle anything outside the expected lance format.  changing it would be a moderate undertaking and as persistent opfor and story arcs are inbound we felt that this would require more effort than it is worth.

**Comment by bandildo** [comment]
Understood.

The only purpose of turning the "L" "UL" "M", etc into a 0-3 number, is to aggregated them. I believe this is to make it more user-friendly when writting the atbconfig. So you could either choose A for assault or SH for super-heavy, that MekHQ will consider it the same. Those letters are only chosen like so to be human-readable. In theory, it should work if you choose any random sequence of letters. It works as an ID for aggregation.

We could easily change to something that does not aggregate, being it simply an ID string. However, this could break some configurations that already take into account this aggregation, which then leads us to the important question: are we okay with that?

In other words, if you use a atbconfig that has both L and UL and you expect them to be aggregated, then that wouldn't work anymore. But the default doesn't have any of these cases, so the vanilla MekHQ would work perfectly fine.

**Comment by bczretribution** [comment]
as you said anyone using things like UL are using a custom to begin with, but the code doesn't implement that as they want anyway, so fixing the code would just mean they would have to modify their custom, which they have already shown they can do.  so i don't see an issue with "breaking" things as its improving implementation and increasing capability.  but then im not part of the dev team, just a guy with ideas.

**Comment by neoancient** [comment]
No, the purpose of turning the String designations of the weight classes into a value [0..3] is that those are the values used by the unit generation system.

**Comment by bandildo** [comment]
@neoancient I don't have the code here, but isn't that only in the entry values of the `botLance` node? I interpreted that the entry values of the `botForce` can be anything that matches as a key for the `botLance`. Being then the values of `botLance` what gets used by the unit generation system.

**Comment by bandildo** [comment]
Okay, the first level (botForce key) should have "L", "M", "H" and "A" like you described, same thing as the lowest level (botLance value). However, the code is being reutilized in the two middle levels (botForce value get a botLance key), creating this issue. Since they just need to correspond a botForce value to a botLance key, it doesn't matter the letters used, but the code re-utilization is creating the constrain. If I understood correctly, of course.

I'll look into it.

**Comment by neoancient** [comment]
None of the code creates this issue. The code does exactly what it was designed to do. The issue is that the proposal does not fit into the parameters of the design, and to accommodate the proposal requires fairly modest change in the design.

**Comment by bandildo** [comment]
I've been looking into this and now understand what you meant.

From what I understood, there are three cases where some special attention is needed:

**Reinforcements**:
I've added an attribute to the `botLance` weighed tables, so you can freely define what is used for reinforcements:

```
			<weightedTable weightClass="L" reinforcementClass="L">
				<entry weight="2">LLLM</entry>
				<entry weight="1">LLMM</entry>
			</weightedTable>
```

**Base Attack Builtin Scenario**:
According to the code, _"Ally deploys 2 **lances** of a lighter weight class than the player, minimum light.". My idea would be to pick a bot **force** of the lighter weight class (since bot forces are divided in LMHA) and then pick a random lance from that force.


**Using dropships**:
With this one I'm a bit more ignorant. Found in `AtBScenario.java`:

```
        if (campaign.getCampaignOptions().getUseDropShips()) {
        	if (canAddDropShips()) {
                for (int i = 0; i < Compute.d6() - 3; i++) {
                    addLance(enemyEntities,
                            getContract(campaign).getEnemyCode(),
                            getContract(campaign).getEnemySkill(),
                            getContract(campaign).getEnemyQuality(),
                            UnitMarket.getRandomWeight(UnitType.MEK, getContract(campaign).getEnemyCode(), campaign.getCampaignOptions().getRegionalMechVariations()),
                            EntityWeightClass.WEIGHT_ASSAULT,
                            campaign);
                }
            } 
```

Which will basically get a random weigh based on regional variations:

```
	public static int getRegionalMechWeight(String faction) {
		int roll = Compute.randomInt(100);
		if (faction.equals("DC")) {
			if (roll < 40) return EntityWeightClass.WEIGHT_LIGHT;
			if (roll < 60) return EntityWeightClass.WEIGHT_MEDIUM;
			if (roll < 90) return EntityWeightClass.WEIGHT_HEAVY;
			return EntityWeightClass.WEIGHT_ASSAULT;
		} 
         }
```

Which it seems the solution would be to implement something similar to the previous point: get a random lance from a force of given weight.

I'd really appreciate feedback and opinion from more knowledgeable members on how to tackle this.

**Comment by sixlettervariables** [comment]
@NickAragua when you have time, could you chime in with whether or not your new scenario generation improvements can cover this request?

**Comment by NickAragua** [comment]
Yes, although I've got a ways to go before the whole thing comes out. The Scenario Template Generator does limit enemy force size based on BV, however.

**Comment by bczretribution** [comment]
the original intent of my project was to act as a fill-in until nick finished his awesome plans.  using his project to implement this one would be moving backward in my mind.  I consider this project abandoned.

**Comment by sixlettervariables** [comment]
Okay thank you @bczretribution
